pkg/cloud/libvirt/actuators/machine/utils/cloudinit.go

```diff
@@ -12,30 +20,63 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// Max firmware config string is set to 1024.
-// Round it to 1020 for some extra waste space.
-const maxFirmwareCfgChunkSize = 1020
+func getCloudInitVolumeName(volumeName string) string {
+	return fmt.Sprintf("%v_cloud-init", volumeName)
+}
+
+func EnsureCloudInitVolumeIsDeleted(name string, client *Client) error {
+	return EnsureVolumeIsDeleted(getCloudInitVolumeName(name), client)
+}
 
-func breakUserDataIntoChunks(userData string, chunkMaxSize int) []string {
-	userDataLen := len(userData)
-	chunks := userDataLen / chunkMaxSize
+func setCloudInit(domainDef *libvirtxml.Domain, client *Client, cloudInit *providerconfigv1.CloudInit, kubeClient kubernetes.Interface, machineNamespace, volumeName, poolName string) error {
 
-	b := make([]string, chunks)
-	for i := 0; i < chunks; i++ {
-		b[i] = userData[i*chunkMaxSize : (i+1)*chunkMaxSize]
+	// At least user data or ssh access needs to be set to create the cloud init
+	if cloudInit.UserDataSecret == "" && !cloudInit.SSHAccess {
+		return nil
+	}
+
+	// default to bash noop
+	userDataSecret := []byte(":")
+	if cloudInit.UserDataSecret != "" {
+		secret, err := kubeClient.CoreV1().Secrets(machineNamespace).Get(cloudInit.UserDataSecret, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("can not retrieve user data secret '%v/%v' when constructing cloud init volume: %v", machineNamespace, cloudInit.UserDataSecret, err)
+		}
+		ok := false
+		userDataSecret, ok = secret.Data["userData"]
+		if !ok {
+			return fmt.Errorf("can not retrieve user data secret '%v/%v' when constructing cloud init volume: key 'data' not found in the secret", machineNamespace, cloudInit.UserDataSecret)
+		}
 	}
 
-	if userDataLen%chunkMaxSize > 0 {
-		b = append(b, userData[chunks*chunkMaxSize:userDataLen])
+	userData, err := renderCloudInitStr(userDataSecret, cloudInit.SSHAccess)
+	if err != nil {
+		return fmt.Errorf("can not render cloud init user-data: %v", err)
 	}
-	return b
-}
 
-func setCloudInit(domainDef *libvirtxml.Domain, cloudInit *providerconfigv1.CloudInit, kubeClient kubernetes.Interface, machineNamespace string) error {
-	if cloudInit.ISOImagePath == "" {
-		return fmt.Errorf("error setting cloud-init, ISO image path is empty")
+	metaData, err := renderMetaDataStr(volumeName)
+	if err != nil {
+		return fmt.Errorf("can not render cloud init meta-data: %v", err)
+	}
+
+	cloudInitISOName := getCloudInitVolumeName(volumeName)
+
+	cloudInitDef := newCloudInitDef()
+	cloudInitDef.UserData = string(userData)
+	cloudInitDef.MetaData = string(metaData)
+	cloudInitDef.Name = cloudInitISOName
+	cloudInitDef.PoolName = poolName
+
+	log.Printf("[INFO] cloudInitDef: %+v", cloudInitDef)
```

Comment:
aren't we using glog for logging now?

Reply:
`log` package is used extensively by the utils so we need to change it at multiple places => follow up PR will be more suitable for this change.

Reply:
yep we want to change across all the library in a different PR, so no need to add new uses for it here,  please let's don't do it in this PR or any new one